### PR TITLE
mrc-3516 change config to add file store

### PR DESCRIPTION
**Pull request by hillalex**
Allows for the adding of a file store

### CODE REVIEW

File: R/root.R [hunk 1]
```diff
@@ -114,6 +114,15 @@ outpack_root <- R6::R6Class(
       private$index_data
     },
 
+    add_file_store = function() {
```
Comment (richfitz) [inline R/root.R:117]: This is I think related to the validation problem (mrc-3052) - if a file is missing or modified here we'll fail and we'll end up with an inconsistent store (a test to verify the behaviour here would be good). Race conditions mean that doing validation-then-import is probably not ideal but it's likely there will be some shared logic.

Probably a reasonable way for this to work would be to *try* and import a packet from its unpacked data and then if that fails see if we can reconstruct it from other files we have kicking about. So if the user has deleted the file data.csv used as an input but we have other copies of that that *are* ok we should just stick one of those in the store. We might already have an equivalent file in the store. When looking through the VIMC and covid stores, there are many duplicated files due to rerunning of reports with the same inputs, and the way that outputs become inputs.

File: R/root.R [hunk 2]
```diff
@@ -114,6 +114,15 @@ outpack_root <- R6::R6Class(
...
+      self$files <- file_store$new(file.path(self$path, ".outpack", "files"))
+      invisible(lapply(self$index()$metadata, function(meta) {
```
Comment (richfitz) [inline R/root.R]: This is going to loop over all packets that the current repo knows about at all, not the ones that are unpacked - we only want the latter. See `index$unpacked$packet` for a list of ids that we should consider

File: R/config.R [hunk 3]
```diff
@@ -84,17 +84,17 @@ config_set_use_file_store <- function(value, root) {
...
-  }
-
   if (!value) {
     if (is.null(config$core$path_archive)) {
       stop("if 'path_archive' is NULL, then 'use_file_store' must be TRUE")
     }
     root$remove_file_store()
   }
 
+  if (value) {
```
Comment (richfitz) [inline R/config.R:94]: This probably wants to be in tryCatch so that if it fails we don't add half a store?

File: R/root.R [hunk 4]
```diff
@@ -114,6 +114,15 @@ outpack_root <- R6::R6Class(
...
+        src <- file.path(self$path, self$config$core$path_archive,
+                         meta$name, meta$id)
+        file_import_store(self, src, meta$files$path, meta$files$hash)
```
Comment (richfitz) [inline R/root.R]: Given the behaviour of `file_import_store` we should filter this down to hashes we don't yet know about. Alternatively, change `file_import_store` to do an "if not exists" check, which is probably tidier...

**Review by richfitz — CHANGES_REQUESTED** [review]

File: R/root.R (hunk 4 above)
Comment (hillalex) [reply R/root.R]: There's already an if not exists check in the file store `put` method: https://github.com/mrc-ide/outpack/blob/main/R/store.R#L50

File: R/root.R (hunk 1 above)
Comment (hillalex) [reply R/root.R:117]: Looks like there is already a function to look up paths from hashes, so using that: https://github.com/mrc-ide/outpack/pull/22/commits/fa0e0323bd8b35da288dcb0e69bd2671b8e3e2ea

**Review by richfitz — APPROVED** [review]

**Pull request merged**